runtime/bcutil/ROMClassCreationContext.hpp

```diff
@@ -293,7 +290,35 @@ class ROMClassCreationContext
 		}
 	}
 
-	void setIsLambda(bool isLambda) { _isLambda = isLambda; }
+	bool isUnsafeClassSharingEnabled() const {
+		/*
+		 * The command line option -XX:[+/-]ShareUnsafeClasses, combined with -XX:[+/-]ShareAnonymousClasses will have 4 different behaviours:
+		 * 1. -XX:+ShareAnonymousClasses -XX:+ShareUnsafeClasses, this will share all unsafe classes
+		 * 2. -XX:+ShareAnonymousClasses -XX:-ShareUnsafeClasses, this will only share anonymous classes and not other unsafe classes
+		 * 3. -XX:-ShareAnonymousClasses -XX:+ShareUnsafeClasses, this will only share unsafe classes that are not anonymous
+		 * 4. -XX:-ShareAnonymousClasses -XX:-ShareUnsafeClasses, this will share neither.
+		 * The current default behavior is the 1st option.
+		 */
+		bool ret = false;
+		if (isClassAnon()) {
+			/* if the class is anonymous, no need to check if ShareAnonymousClasses is enabled because the class loader flag
+			 * is already set accordingly (in the function Java_sun_misc_Unsafe_defineAnonymousClass in sun_misc_Unsafe.cpp)
+			 * and it is being checked in isRomClassShareable() before this function is called. So instead we can assert if
+			 * ShareAnonymousClasses is disabled
+			 */
+			if ((NULL != _javaVM) && (NULL != _javaVM->sharedClassConfig)) {
+				Trc_BCU_Assert_True(J9_ARE_ALL_BITS_SET(_javaVM->sharedClassConfig->runtimeFlags, J9SHR_RUNTIMEFLAG_ENABLE_SHAREUNSAFECLASSES));
+			}
+			ret = true;
+		} else if ((NULL != _javaVM)
+				   && (NULL != _javaVM->sharedClassConfig)
+				   && J9_ARE_ALL_BITS_SET(_javaVM->sharedClassConfig->runtimeFlags, J9SHR_RUNTIMEFLAG_ENABLE_SHAREUNSAFECLASSES)
+		) {
+			/* if the class is unsafe but not anonymous, we need to check if ShareUnsafeClasses is enabled */
+			ret = true;
+		}
+		return ret;
+	}
```

Comment:
Something like this localizes the bit checks and makes the decision logic easier to see:
```
	bool isUnsafeClassSharingEnabled() const {
		/*
		 * The command line option -XX:[+/-]ShareUnsafeClasses, combined with -XX:[+/-]ShareAnonymousClasses will have 4 different behaviours:
		 * 1. -XX:+ShareAnonymousClasses -XX:+ShareUnsafeClasses, this will share all unsafe classes
		 * 2. -XX:+ShareAnonymousClasses -XX:-ShareUnsafeClasses, this will only share anonymous classes and not other unsafe classes
		 * 3. -XX:-ShareAnonymousClasses -XX:+ShareUnsafeClasses, this will only share unsafe classes that are not anonymous
		 * 4. -XX:-ShareAnonymousClasses -XX:-ShareUnsafeClasses, this will share neither.
		 * The current default behavior is the 1st option.
		 */
		bool isEnabled = false;

		bool isAnonDefineClassSharingEnabled = false;
		bool isUnsafeDefineClassSharingEnabled = false;
		if ((NULL != _javaVM)
		&& (NULL != _javaVM->sharedClassConfig)
		) {
			isAnonDefineClassSharingEnabled = J9_ARE_ALL_BITS_SET(_javaVM->sharedClassConfig->runtimeFlags, J9SHR_RUNTIMEFLAG_ENABLE_SHAREANONYMOUSCLASSES);
			isUnsafeSharingEnabled = J9_ARE_ALL_BITS_SET(_javaVM->sharedClassConfig->runtimeFlags, J9SHR_RUNTIMEFLAG_ENABLE_SHAREUNSAFECLASSES);
		}

		if (isClassAnon()) {
			/* if the class is anonymous, no need to check if ShareAnonymousClasses is enabled because the class loader flag
			 * is already set accordingly (in the function Java_sun_misc_Unsafe_defineAnonymousClass in sun_misc_Unsafe.cpp)
			 * and it is being checked in isRomClassShareable() before this function is called. So instead we can assert if
			 * ShareAnonymousClasses is disabled
			 */
			Trc_BCU_Assert_True(J9_ARE_ALL_BITS_SET(isAnonDefineClassSharingEnabled);
			isEnabled = true;
		} else if (isUnsafeDefineClassSharingEnabled) { 
			/* Classes loaded by Unsafe.defineClass */
			isEnabled = true;
		}
		return isEnabled;
	}
```